### ISSUE TITLE
Problems with libpng version used for fluid vs. application code

### ISSUE THREAD

**Issue by pdhahn** [fltk/fltk#425]
As of commit 5a8b9d9130d9c9bd0adeca33ac051eb9249953fb (dated April 3?) in 1.4 I now have a problem with warnings emitted when I start fluid:

`Fl_File_Icon::load(): Unable to load icon file "/usr/share/icons/crystalsvg/16x16/actions/help.png"`

and many more like that. Note I verified all of these `.png` files exist, so I believe the message is related to the version of libpng being employed.

I am using CentOS 6 (latest) on a Dell workstation. I only build FLTK 1.4 static libs.

In my cmake config, I have a path to a fairly modern version of libpng in PNG_LIBRARY_RELEASE because I need it to link with my application code. This version of libpng is from 2021.

However, because my system is old, and the icon files referenced above are dated something like 2005, I also have the cmake config flag OPTION_USE_SYSTEM_LIBPNG set. The system version of libpng is from 2015.

This dual-version of libpng arrangement worked fine for me until the commit version referenced above. I need some way to continue with this version dichotomy if possible, please.

**Comment by Albrecht-S** [comment]
@pdhahn Thanks for your report. I just fixed an issue accidentally introduced by the mentioned commit (5a8b9d913) that is very likely the cause of your issue. Can you please pull the latest Git master (at least commit 8bd67f40daa0777de3a2ee150164f0c4a9f44efc) and test if your issue persists?

If the issue is resolved, please confirm. But please see below for some questions about your build system which I'd appreciate to get answered. [1]

If it is not resolved, please report the details how exactly you setup your CMake config ("a path to a fairly modern version of libpng in PNG_LIBRARY_RELEASE") so I can test and see what this may cause. Please tell us also which concrete libpng versions you have located in "PNG_LIBRARY_RELEASE" and the system png version. The bundled libpng version is 1.6.37.

BTW: this all should not be affected by the date of the icon (I assume you mean the .png) files. Every PNG library version should be able to open these files. But thanks in advance for any insights if this is not the case.

FYI: since Nov 14, 2021 (commit 46cbc31de) we're using "prefixed" libpng, libjpeg, and zlib versions in our bundled libraries which are used if you set `OPTION_USE_SYSTEM_{LIBJPEG|LIBPNG|ZLIB}=OFF`. This should only affect the internal usage of these libraries in FLTK and not in your application if you link your application with another version of libpng. Since we have the prefixed libraries the "safest" approach for using any other version of libpng in your program would be to set all these above mentioned options **OFF** and use whatever you need in your own build.

[1] I'm very much interested in your setup so I can see how your application and our FLTK build interact. Are you building FLTK separately or are you integrating the FLTK build by any means (add_subdirectory or ...?) in your application build? Other users may be using a similar setup and every information could help to improve FLTK. Thanks in advance.

**Comment by pdhahn** [comment]
OK I just now did git pull on master and rebuilt. Yes, the problem is gone. :-)

Yes, I build FLTK static libs only directly from cloned source and my application code links with those `.a`'s. I have a `build` subdirectory under the cloned source and I do my `cmake` and `make` therein.

As for my libpng versions (after resolving softlinks):

- installed on my system:  `/usr/lib64/libpng12.so.0.49.0`
- my custom installation: `.../lib64/libpng16.so.16.38.git`

When I do `ldd .../fltk-clone/build/bin/fluid`, I actually see my custom version, so my guesswork about fluid linking with the system version was wrong:

`libpng16.so.16 => .../lib64/libpng16.so.16`

Also, I unchecked the OPTION_USE_SYSTEM_LIBPNG as you suggested. My application code when linked to the resulting FLTK seems fine and starting fluid is clean. Moreover, `ldd fluid` still reports the same thing.

FYI here is the report from cmake-gui as to configuration after unchcking OPTION_USE_SYSTEM_LIBPNG:

> Configuration Summary for FLTK 1.4.0 generated by CMake 3.19.0 --
>
> -- The following OPTIONAL packages have been found:
>
> * Doxygen
> * LATEX
> * ZLIB
> * JPEG
>
> Static libraries will be built in .../fltk-clone/build/lib
> Shared libraries will not be built (set OPTION_BUILD_SHARED_LIBS=ON to build)
> Test programs    will be built in .../fltk-clone/build/bin/test
> Example programs will be built in .../fltk-clone/build/bin/examples
> Image Libraries: JPEG = System
>                 PNG  = Builtin
>                 ZLIB = System
> Use Wayland:     No
> Use Pango:       No

Thanks!

**Comment by Albrecht-S** [comment]
@pdhahn Thank you for confirmation and the additional information.

Rearding libpng and zlib: I'm thinking about forcing both libs to be built from the bundled version **_if any one of them uses the bundled lib_**. This is not yet done but I recommend using either both bundled or both system versions because libpng depends on zlib and if you use the internal libpng and the (shared) system zlib then this can be a version conflict that might appear at runtime (by a runtime check in libpng if this exists - I'm not sure) or it can lead to unnoticed runtime problems.

libjpeg is independent of the other two libs and can be chosen as you like.

**Comment by Albrecht-S** [comment]
@pdhahn As for your "special version" of libpng: I'm pretty sure I worked out a way to build zlib, all FLTK image libraries and (for our needs) libtiff as external libs that would be used on all platforms instead of linking any of these as system libs. Linking all these libs as static libs made our application independent of any version of these on the target platform. I don't remember exactly how this was done (and I don't have access to this app anymore) but I'm pretty sure we used CMake to build FLTK and our application.

I assume that you would need to set library locations (such as PNG_LIBRARY_RELEASE or similar) in your own build and set `OPTION_USE_SYSTEM_{LIBJPEG|LIBPNG|ZLIB}=**ON**` for FLTK (and your app) to find the same pre-built image libraries to use in your app.

Caveat: I'm currently working on an issue that using the bundled image libs combined with Pango and/or Cairo leads to conflicting include directories which may affect the build with undefined outcome (either failed build or maybe runtime issues). This is still under investigation.

**Comment by Albrecht-S** [comment]
Note: although the concrete issue is resolved I reopened this issue for myself and others to follow the further discussion.

> I am not clear on how the bundled libpng actually gets used concretely in FLTK code, while still allowing for a dynamic reference to an external version of libpng (i.e., as with fluid).

Well, it's complicated and it's confusing, and complicated, and confusing, ... ;-)

What you see in `ldd fluid` (as you wrote above) may look different than it really is. It was much easier before we used xft, Cairo, Pango, etc.. See the (edited) build log below (I'm using `ninja` to build):
```
$ cmake -G Ninja \
>   -DCMAKE_BUILD_TYPE:STRING=Debug \
>   -DOPTION_USE_XFT:BOOL=OFF \
>   -DOPTION_USE_SYSTEM_LIBJPEG:BOOL=OFF \
>   -DOPTION_USE_SYSTEM_LIBPNG:BOOL=OFF \
>   -DOPTION_USE_SYSTEM_ZLIB:BOOL=OFF \
>   ..
-- The C compiler identification is GNU 9.4.0
[...]
-- Configuration Summary for FLTK 1.4.0 generated by CMake 3.16.3 --

-- The following OPTIONAL packages have been found:

 * Doxygen
 * LATEX

-- Static libraries will be built in /git/fltk/master/build-test/lib
-- Shared libraries will not be built (set OPTION_BUILD_SHARED_LIBS=ON to build)
-- Test programs    will be built in /git/fltk/master/build-test/bin/test
-- Example programs will not be built (set FLTK_BUILD_EXAMPLES=ON to build)
-- Image Libraries: JPEG = Builtin
--                  PNG  = Builtin
--                  ZLIB = Builtin
-- Use Wayland:     No
-- Use Pango:       No
-- Use Xft:         No
-- Cairo support:   No

-- End of Configuration Summary --

-- Configuring done
-- Generating done
-- Build files have been written to: /git/fltk/master/build-test
$ ninja
[555/555] Linking CXX executable bin/test/resize
$ ldd bin/fluid
	linux-vdso.so.1 (0x00007fffe3f88000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f974c5d1000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f974c5ae000)
	libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007f974c471000)
	libXinerama.so.1 => /usr/lib/x86_64-linux-gnu/libXinerama.so.1 (0x00007f974c46c000)
	libXfixes.so.3 => /usr/lib/x86_64-linux-gnu/libXfixes.so.3 (0x00007f974c464000)
	libXcursor.so.1 => /usr/lib/x86_64-linux-gnu/libXcursor.so.1 (0x00007f974c457000)
	libXrender.so.1 => /usr/lib/x86_64-linux-gnu/libXrender.so.1 (0x00007f974c24b000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f974c069000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f974bf1a000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f974beff000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f974bd0d000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f974c942000)
	libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f974bce3000)
	libXext.so.6 => /usr/lib/x86_64-linux-gnu/libXext.so.6 (0x00007f974bccc000)
	libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6 (0x00007f974bcc6000)
	libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f974bcbe000)
	libbsd.so.0 => /usr/lib/x86_64-linux-gnu/libbsd.so.0 (0x00007f974bca4000)
$ 
```
This log shows that `fluid` itself does NOT link to any external (shared) libpng version (neither bundled nor system nor other referenced version) if it is built w/o xft, cairo, and pango support with the bundled libpng as a static library. The way this is achieved is by specifying for instance '-I ../png' (early) on the commandline so all references to libpng headers are resolved by the bundled libpng version and a direct reference to libfltk_png.a on the linker commandline.

The problem is that xft, cairo, and pango depend on libpng, which means as soon as you link with one of these as shared libraries, the resultant executable will also show a dependency on this shared library. This can be shown easily by replacing `-DOPTION_USE_XFT:BOOL=OFF` with `-DOPTION_USE_XFT:BOOL=ON` in the example above. If you do this you'll see. This was one of the main reasons why we switched to the prefixed bundled image libraries (but to explain the details would be even more complicated).

You can verify that libxft.so depends on libpng by running `ldd` on it:
```
$ ldd /usr/lib/x86_64-linux-gnu/libXft.so.2
	linux-vdso.so.1 (0x00007ffed0fe2000)
...
	libpng16.so.16 => /usr/lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f34109d6000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f34109ba000)
```

So, in theory, the FLTK libs can be compiled using their own internal bundled image libs (linked statically) while still allowing libXft, libcairo, and libpango to be linked with their own shared library version of libpng. However, it's way more complicated in practice - and this is what I'm working on.

In practice we need to specify the relevant libcairo and libpango include directories to compile the cairo and pango code. Currently we use only one common set of `-I` flags for all code. This leads to more than one '-I<path-to-libpng>' include path in the compile command and only one (the first appearance) can win. This makes it possible that we compile "png" code (i.e. code that uses png structs) with the wrong version of the libpng headers, i.e. with the wrong ABI - with unpredictable results.

The solution I'm investigating is to selectively use "the right" `-I` flags depending on the code we compile, something like
- compiling `libfltk_image` code uses the `-I` flag for the bundled libpng
- compiling FLTK drawing code uses the `-I` flag(s) for Pango and Cairo etc.

I believe that this is doable with CMake using the correct flags per build unit. I'm not a configure/make expert and I'm not sure that I can do the same for configure/make.

Anyway, once this is done correctly (and it's not done yet), you should be able to compile and link your own source code with your own libpng version, but if yours is different than the system libpng version pulled in by the shared libXft and/or libcairo versions you may also need to "prefix" your own libpng version. Otherwise you might see the same issues we had with FLTK and libXft/libpango which led to application crashes and other strange effects. In this case I would suggest to let FLTK use the system libpng version if possible which would simplify things.

The best you can do is to use the system version of all image libraries and not to mix two different versions. What we're doing with FLTK (allowing the bundled image libs) has been done before we used Xft and Pango which included their own libpng dependencies and led to these issues and the consequence to use image lib prefixing.

I'm not sure if I could explain everything sufficiently but I hope you got an impression of how complicated this can be.

**Comment by pdhahn** [comment]
@[Albrecht-S](https://github.com/Albrecht-S) Thanks for informing me about the challenges involved with knitting the various dependencies together. I can now better appreciate what you are having to deal with in this area and the necessity for even some tricky things. :-)

> Anyway, once this is done correctly (and it's not done yet), you should be able to compile and link your own source code with your own libpng version, but if yours is different than the system libpng version pulled in by the shared libXft and/or libcairo versions you may also need to "prefix" your own libpng version.<

Excellent -- I will look forward to that!

**Comment by pdhahn** [comment]
BTW here's the error message from cmake-gui when I check the OPTION_CAIRO flag  (checking OPTION_CAIRO_EXT seems to make no improvement) then do configure:

> Checking for one of the modules 'cairo'
*** Cairo was requested but not found - please check your cairo installation
***   or disable options OPTION_CAIRO and OPTION_CAIRO_EXT.
CMake Error at CMake/options.cmake:369 (message):
  *** Terminating: missing Cairo libs or headers.
Call Stack (most recent call first):
  CMakeLists.txt:76 (include)

although I have a version of cairo installed on my system thus:

`488 -rwxr-xr-x  1 root root 497448 Jun 22  2015 /usr/lib64/libcairo.so.2.10800.8`

along with the associated "developer" header files, e.g.,

`80 -rw-r--r--. 1 root root 79860 Jun 22  2015 /usr/include/cairo/cairo.h`
`...`

As I mentioned above, there was a point in time before which I was in fact able to check the OPTION_CAIRO flag and it (the configure/build step) for FLTK worked, which built the `cairo_test` executable that I could run successfully.

**Comment by Albrecht-S** [comment]
OK, sorry, this Cairo stuff is now going way too much off topic, particularly because the issue is already fixed and we are already discussing off-topic libpng stuff here. (FTR: I'm making progress in rewriting the image lib stuff...)

I'm ignoring everything related to Cairo configuration in this thread to avoid confusion and to finish (close) this issue soon.

I know that we're using `pkg-config` to find the cairo libs and flags nowadays, maybe this was different some time ago, but if, then it was some "hand made" stuff that was replaced with `pkg-config` for a good reason, so you should probably check if pkg-config can find the cairo libs. That said, if you need help with configuration, please ask your question in fltk.general first. Thanks.

Standard reply follows:

> This is the wrong place to ask questions about building or running programs with FLTK (user support). FLTK "issues" are reserved for bug reports and requests for enhancements, for further information please visit https://www.fltk.org/bugs.php.
>
>Please ask user questions in our user group [fltk.general](https://groups.google.com/forum/#!forum/fltkgeneral). Your question will have a wider audience and you will likely get more and quicker replies.
>
> Please include the system/platform (Windows, Linux, macOS), the build system (Visual Studio, CMake, Ninja), compiler (gcc, clang), compiler version, and build error messages in text form (copy and paste preferred). TIA.

**Comment by pdhahn** [comment]
OK aorry about the Cairo stuff.

Back to png, I notice if I uncheck "Use system png', then build static, something still seems awry (maybe?) when I invoke the resulting `fltk-config`:
```
fltk-config --use-forms --use-images --use-gl --ldstaticflags
```
produces 

```
fltk/lib64/libfltk_images.a -ljpeg -lpng -lz  fltk/lib64/libfltk_gl.a -lGLU -lGL fltk/lib64/libfltk_forms.a fltk/lib64/libfltk.a -lm -lX11 -lXext -lpthread -lXinerama -lXfixes -lXcursor -lXft -lXrender -lm -lfontconfig -ldl
```
Notice there is no `fltk/lib64/libfltk_png.a`, which is generated by the build, and `fltk/lib64/libfltk_images.a` contains references to symbols like `fltk_png_init_io`, defined in the former.

Perhaps I am not understanding things properly, but I would expect `fltk-config` to list the `libfltk_png.a` after `libfltk_images.a` to resolve the FLTK-specific undefined symbols. Please advise.

**Comment by Albrecht-S** [comment]
@pdhahn Thanks for this report as well. I can confirm that `fltk-config` produces wrong references to image libraries as it is today in git master.

The generation of `fltk-config` by `CMake` still has a lot of deficiencies which should be handled in issue #129. This issue got a long discussion about related (or not so much related) topics but I added your point as another comment so it will not be forgotten (this issue is meant to be worked on by me definitely before the release of 1.4.0).

**Comment by Albrecht-S** [comment]
I believe that this issue is resolved and can be closed. The latest discussions were about OT libpng stuff anyway.

Meanwhile we have a new feature in CMake and configure that makes sure that both libpng and zlib are always configured as either both the system versions or both the bundled (and prefixed) versions. This ensures that we don't "mix" libraries of incompatible versions.

The problem of `fltk-config` not emitting the correct image libraries will be handled in issue #129.

I'm closing this issue now. If you have any remaining problems please open another issue.
